File: src/worker/recorder.ts
import { createWriteStream, WriteStream } from "fs";
import { EventEmitter } from "events";

export interface CastHeader {
  version: 2;
  width: number;
  height: number;
  timestamp: number;
  title?: string;
  env?: Record<string, string>;
}

export interface CastEvent {
  time: number;
  type: "o" | "i";
  data: string;
}

/**
 * Records a terminal session in asciinema v2 cast format.
 *
 * Usage:
 *   const rec = new AsciinemaRecorder({ outputPath, cols, rows });
 *   rec.start();
 *   // ... later ...
 *   rec.write(chunk);   // called with each PTY data chunk
 *   rec.stop();
 *   await rec.flush();
 */
export class AsciinemaRecorder extends EventEmitter {
  private _stream: WriteStream | null = null;
  private _startTime: number = 0;
  private _events: CastEvent[] = [];
  private readonly outputPath: string;
  private readonly cols: number;
  private readonly rows: number;
  private readonly title: string;

  constructor(options: {
    outputPath: string;
    cols: number;
    rows: number;
    title?: string;
  }) {
    super();
    this.outputPath = options.outputPath;
    this.cols = options.cols;
    this.rows = options.rows;
    this.title = options.title ?? "TUIcast Recording";
  }

  /** Begin the recording session. */
  start(): void {
    this._startTime = Date.now();
    this._events = [];
    this._stream = createWriteStream(this.outputPath, { encoding: "utf8" });

    const header: CastHeader = {
      version: 2,
      width: this.cols,
      height: this.rows,
      timestamp: Math.floor(this._startTime / 1000),
      title: this.title,
      env: { TERM: "xterm-256color", SHELL: "/bin/bash" },
    };
    this._stream.write(JSON.stringify(header) + "\n");
  }

  /**
   * Record a chunk of output received from the PTY.
   * @param data – Raw terminal output (may contain ANSI sequences).
   */
  write(data: string): void {
    if (!this._stream) return;
    const elapsed = (Date.now() - this._startTime) / 1000;
    const event: CastEvent = { time: elapsed, type: "o", data };
    this._events.push(event);
    // asciinema v2 format: each event is a JSON array on its own line.
    this._stream.write(JSON.stringify([event.time, event.type, event.data]) + "\n");
  }

  /** Finish recording and close the output stream. */
  async stop(): Promise<void> {
    if (!this._stream) return;
    await new Promise<void>((resolve, reject) => {
      this._stream!.end((err: Error | null | undefined) => {
        if (err) reject(err);
        else resolve();
      });
    });
    this._stream = null;
  }

  /** Duration of the recording in seconds. */
  get durationSeconds(): number {
    if (this._events.length === 0) return 0;
    return this._events[this._events.length - 1].time;
  }

  /** Number of recorded events. */
  get eventCount(): number {
    return this._events.length;
  }
}

File: pkg/recorder/recorder.go
// Package recorder writes asciinema v2 cast files.
package recorder

import (
	"encoding/json"
	"fmt"
	"io"
	"time"
	"unicode/utf8"
)

const (
	defaultWidth  = 120
	defaultHeight = 30
)

// Config describes an asciinema recording.
type Config struct {
	Width     int
	Height    int
	Timestamp time.Time
	Title     string
	Env       map[string]string
	Clock     Clock
}

// Recorder writes asciinema v2 output events as they arrive.
type Recorder struct {
	writer  io.Writer
	clock   Clock
	pending []byte
	closed  bool
}

type header struct {
	Version   int               `json:"version"`
	Width     int               `json:"width"`
	Height    int               `json:"height"`
	Timestamp int64             `json:"timestamp"`
	Title     string            `json:"title,omitempty"`
	Env       map[string]string `json:"env,omitempty"`
}

// New creates a recorder and writes the asciinema v2 header immediately.
func New(writer io.Writer, config Config) (*Recorder, error) {
	config = normalizeConfig(config)
	line, err := json.Marshal(header{
		Version:   2,
		Width:     config.Width,
		Height:    config.Height,
		Timestamp: config.Timestamp.Unix(),
		Title:     config.Title,
		Env:       config.Env,
	})
	if err != nil {
		return nil, fmt.Errorf("marshal cast header: %w", err)
	}

	if _, err := writer.Write(append(line, '\n')); err != nil {
		return nil, fmt.Errorf("write cast header: %w", err)
	}

	return &Recorder{
		writer: writer,
		clock:  config.Clock,
	}, nil
}

// Write records PTY output bytes, buffering incomplete UTF-8 suffixes.
func (r *Recorder) Write(p []byte) (int, error) {
	if r.closed {
		return 0, fmt.Errorf("recorder is closed")
	}

	data := append(r.pending, p...)
	complete, pending := splitCompleteUTF8(data)
	r.pending = append(r.pending[:0], pending...)

	if len(complete) == 0 {
		return len(p), nil
	}

	if err := r.writeOutput(string(complete)); err != nil {
		return 0, err
	}

	return len(p), nil
}

// Close flushes any pending bytes and prevents further writes.
func (r *Recorder) Close() error {
	if r.closed {
		return nil
	}

	r.closed = true
	if len(r.pending) == 0 {
		return nil
	}

	pending := string(r.pending)
	r.pending = nil

	return r.writeOutput(pending)
}

func normalizeConfig(config Config) Config {
	if config.Width <= 0 {
		config.Width = defaultWidth
	}

	if config.Height <= 0 {
		config.Height = defaultHeight
	}

	if config.Timestamp.IsZero() {
		config.Timestamp = time.Now()
	}

	if config.Clock == nil {
		clock := NewWallClock()
		config.Clock = clock
	}

	return config
}

func (r *Recorder) writeOutput(output string) error {
	event := []any{r.clock.Elapsed().Seconds(), "o", output}
	line, err := json.Marshal(event)
	if err != nil {
		return fmt.Errorf("marshal output event: %w", err)
	}

	if _, err := r.writer.Write(append(line, '\n')); err != nil {
		return fmt.Errorf("write output event: %w", err)
	}

	return nil
}

func splitCompleteUTF8(data []byte) ([]byte, []byte) {
	if len(data) == 0 {
		return nil, nil
	}

	start := len(data) - utf8.UTFMax
	if start < 0 {
		start = 0
	}

	for i := len(data) - 1; i >= start; i-- {
		if data[i] < utf8.RuneSelf {
			return data, nil
		}

		if !utf8.RuneStart(data[i]) {
			continue
		}

		if utf8.FullRune(data[i:]) {
			return data, nil
		}

		return data[:i], data[i:]
	}

	return data, nil
}
